### PR TITLE
Use NonBlocking.ConcurrentDictionary

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- Use [`NonBlocking.ConcurrentDictionary`](https://github.com/VSadov/NonBlocking) from @VSadov where appropriate. Can't changes `AllNodes` in `TrieStore` as we also grab all the internal locks at one point.

#### What types of changes does your code introduce?

- [x] Optimization

#### Requires testing

- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs [hunk 1]
```diff
@@ -354,7 +354,7 @@ private void LogStateOnPrepare()
             {
                 lock (_handlerLock)
                 {
-                    ConcurrentDictionary<long, string> all = new();
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs:357]: This is under lock so doesn't need to be Concurrent

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs [hunk 2]
```diff
@@ -831,7 +831,7 @@ void ClearCommitSetQueue()
                     PruneCache();
                     KeyValuePair<ValueHash256, TrieNode>[] nodesCopy = _dirtyNodes.AllNodes.ToArray();
 
-                    ConcurrentDictionary<ValueHash256, bool> wasPersisted = new();
+                    NonBlocking.ConcurrentDictionary<Hash256AsKey, bool> wasPersisted = new();
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:834]: Use object/pointer sized key or will box it to be pointer sized

**Comment by asdacap** [comment]
Got benchmark? The PatriciaTrie benchmark should touch this on the memory tree store benchmark.

**Comment by benaadams** [comment]
> Got benchmark? The PatriciaTrie benchmark should touch this on the memory tree store benchmark.

Improvement but not a huge change

Before

| Method                          | Mean       | Error    | StdDev   |
|-------------------------------- |-----------:|---------:|---------:|
| Scenarios                       |   105.3 us |  0.56 us |  0.47 us |
| InsertAndHash                   | 2,788.1 us | 17.79 us | 14.85 us |
| InsertAndCommit                 | 5,926.4 us | 58.82 us | 52.14 us |
| ReadWithFullTree                | 2,626.0 us | 13.83 us | 10.80 us |
| ReadWithUncommittedFullTree     |   480.7 us |  1.96 us |  1.64 us |
| ReadWithMemoryTrieStore         | 1,776.2 us | 17.84 us | 16.69 us |
| ReadWithMemoryTrieStoreReadOnly | 2,695.9 us |  8.83 us |  7.37 us |
| ReadAndDeserialize              | 2,634.0 us | 28.22 us | 25.01 us |

After

| Method                          | Mean       | Error     | StdDev    |
|-------------------------------- |-----------:|----------:|----------:|
| Scenarios                       |   103.3 us |   0.91 us |   0.81 us |
| InsertAndHash                   | 2,676.2 us |  27.76 us |  23.18 us |
| InsertAndCommit                 | 5,885.4 us | 112.13 us | 104.89 us |
| ReadWithFullTree                | 2,521.6 us |  37.79 us |  35.35 us |
| ReadWithUncommittedFullTree     |   464.1 us |   3.68 us |   3.45 us |
| ReadWithMemoryTrieStore         | 1,755.0 us |  21.59 us |  20.20 us |
| ReadWithMemoryTrieStoreReadOnly | 2,591.2 us |  12.50 us |  10.44 us |
| ReadAndDeserialize              | 2,543.3 us |  32.45 us |  30.35 us |

**Comment by benaadams** [comment]
Zero processing spikes since switchover, but going to say that is coincidence
image

**Review by asdacap — APPROVED** [review]

**Pull request merged**